File: tif_customization/tif_customization/page/donation_collection_/donation_collection_.py
import json
import calendar
from datetime import date

import frappe
from frappe.utils import flt


@frappe.whitelist()
def get_donation_collection_data(filters=None):
	if isinstance(filters, str):
		filters = json.loads(filters or "{}")
	filters = filters or {}

	from_date = filters.get("from_date")
	to_date = filters.get("to_date")
	period_type = (filters.get("period_type") or "").strip().lower()
	reference_date = filters.get("reference_date") or from_date

	if not from_date or not to_date:
		frappe.throw("From Date and To Date are required.")

	query_from = from_date
	query_to = to_date
	if period_type == "month":
		months = _fiscal_year_month_blocks(reference_date)
		query_from = months[0]["from_date"]
		query_to = months[-1]["to_date"]
	else:
		months = _month_blocks(from_date, to_date)

	rows = frappe.db.sql(
		"""
		SELECT
			COALESCE(d.donor_name, d.donor, 'Walk-in / Unknown') AS donor_name,
			COALESCE(d.donor, '') AS donor_id,
			DATE_FORMAT(d.donation_date, '%%Y-%%m') AS month_key,
			SUM(COALESCE(d.donation_amount, 0)) AS committed_amount,
			SUM(CASE WHEN COALESCE(dt.category, '') = 'Zakat' THEN COALESCE(d.received_amount, 0) ELSE 0 END) AS zakat_fund,
			SUM(CASE WHEN COALESCE(dt.category, '') != 'Zakat' THEN COALESCE(d.received_amount, 0) ELSE 0 END) AS other_collection,
			SUM(COALESCE(d.outstanding_amount, 0)) AS balance_amount
		FROM `tabDonation` d
		LEFT JOIN `tabDonation Type` dt ON dt.name = d.donation_type
		WHERE d.docstatus = 1
		  AND d.donation_date BETWEEN %(from_date)s AND %(to_date)s
		GROUP BY donor_name, donor_id, month_key
		ORDER BY donor_name, month_key
		""",
		{"from_date": query_from, "to_date": query_to},
		as_dict=True,
	)
	commitment_rows = frappe.db.sql(
		"""
		SELECT
			COALESCE(dn.donor_name, dc.parent, 'Walk-in / Unknown') AS donor_name,
			COALESCE(dc.parent, '') AS donor_id,
			SUM(COALESCE(dc.commitment_amount, 0)) AS commitment_amount
		FROM `tabDonor Commitment` dc
		LEFT JOIN `tabDonor` dn ON dn.name = dc.parent
		WHERE COALESCE(dc.status, 'Active') != 'Cancelled'
		  AND dc.from_date <= %(to_date)s
		  AND dc.to_date >= %(from_date)s
		GROUP BY donor_name, donor_id
		ORDER BY donor_name
		""",
		{"from_date": from_date, "to_date": to_date},
		as_dict=True,
	)
	month_keys = [m["key"] for m in months]
	by_donor = {}
	commitment_by_donor = {}

	for c in commitment_rows:
		key = f"{c.get('donor_name')}::{c.get('donor_id')}"
		commitment_by_donor[key] = flt(c.get("commitment_amount"))
	grand = {
		"committed_amount": 0.0,
		"other_collection": 0.0,
		"zakat_fund": 0.0,
		"total_collection": 0.0,
		"balance_amount": 0.0,
		"month_values": {k: 0.0 for k in month_keys},
	}

	for r in rows:
		zakat_fund = flt(r.get("zakat_fund"))
		other_collection = flt(r.get("other_collection"))
		total_collection = other_collection + zakat_fund
		donor_key = f"{r.get('donor_name')}::{r.get('donor_id')}"
		if donor_key not in by_donor:
			by_donor[donor_key] = {
				"donor_name": r.get("donor_name") or "Walk-in / Unknown",
				"donor_id": r.get("donor_id") or "",
				"month_values": {k: 0.0 for k in month_keys},
				"committed_amount": commitment_by_donor.get(donor_key, 0.0),
				"other_collection": 0.0,
				"zakat_fund": 0.0,
				"total_collection": 0.0,
				"balance_amount": 0.0,
			}

		item = by_donor[donor_key]
		month_key = r.get("month_key")
		if month_key in item["month_values"]:
			item["month_values"][month_key] += total_collection
			grand["month_values"][month_key] += total_collection

		item["other_collection"] += other_collection
		item["zakat_fund"] += zakat_fund
		item["total_collection"] += total_collection
		item["balance_amount"] += flt(r.get("balance_amount"))

		grand["other_collection"] += other_collection
		grand["zakat_fund"] += zakat_fund
		grand["total_collection"] += total_collection
		grand["balance_amount"] += flt(r.get("balance_amount"))

	# include donors having commitments but no donations in selected period
	for donor_key, committed_amount in commitment_by_donor.items():
		if donor_key in by_donor:
			continue
		donor_name, donor_id = donor_key.split("::", 1)
		by_donor[donor_key] = {
			"donor_name": donor_name or "Walk-in / Unknown",
			"donor_id": donor_id or "",
			"month_values": {k: 0.0 for k in month_keys},
			"committed_amount": committed_amount,
			"other_collection": 0.0,
			"zakat_fund": 0.0,
			"total_collection": 0.0,
			"balance_amount": 0.0,
		}

	grand["committed_amount"] = sum(flt(d.get("committed_amount")) for d in by_donor.values())

	data = sorted(by_donor.values(), key=lambda d: (d.get("donor_name") or "").lower())
	return {"months": months, "rows": data, "totals": grand}


def _month_blocks(from_date, to_date):
	start = frappe.utils.getdate(from_date)
	end = frappe.utils.getdate(to_date)
	cur = date(start.year, start.month, 1)
	limit = date(end.year, end.month, 1)
	out = []
	while cur <= limit:
		month_last_day = calendar.monthrange(cur.year, cur.month)[1]
		out.append(
			{
				"key": f"{cur.year}-{cur.month:02d}",
				"label": cur.strftime("%B %Y"),
				"from_date": f"{cur.year}-{cur.month:02d}-01",
				"to_date": f"{cur.year}-{cur.month:02d}-{month_last_day:02d}",
			}
		)
		if cur.month == 12:
			cur = date(cur.year + 1, 1, 1)
		else:
			cur = date(cur.year, cur.month + 1, 1)
	return out


def _fiscal_year_month_blocks(reference_date):
	ref = frappe.utils.getdate(reference_date)
	fy_start_year = ref.year if ref.month >= 7 else ref.year - 1
	return _month_blocks(f"{fy_start_year}-07-01", f"{fy_start_year + 1}-06-30")

File: tif_customization/tif_customization/page/donation_collection_/donation_collection_.js
frappe.pages["donation-collection-"].on_page_load = function (wrapper) {
	console.log("[DonationCollection] on_page_load fired");
	const page = frappe.ui.make_app_page({
		parent: wrapper,
		title: "Donation Collection Position",
		single_column: true,
	});

	if (!window.DonationCollectionReport) {
		window.DonationCollectionReport = class DonationCollectionReport {
			constructor(page) {
				this.page = page;
				this.filters = {
					period_type: "half_yearly",
					reference_date: frappe.datetime.get_today(),
					from_date: "",
					to_date: "",
				};
			}

			make() {
				console.log("[DonationCollection] make() start", { filters: this.filters });
				this.set_dates_from_period();
				this.render();
				this.inject_styles();
				this.bind_events();
				this.load_data();
				console.log("[DonationCollection] make() complete", { filters: this.filters });
			}

			render() {
				console.log("[DonationCollection] render() called", { period_type: this.filters.period_type });
				this.page.main.html(`
					<div class="donation-filter-box">
						<div class="row">
							<div class="col-md-3">
								<label>Period Type</label>
								<select id="donation-period-type" class="form-control"></select>
							</div>
							<div class="col-md-3">
								<label>Reference Date</label>
								<input type="date" id="donation-reference-date" class="form-control" value="${this.filters.reference_date}">
							</div>
							<div class="col-md-3">
								<label>From Date</label>
								<input type="date" id="donation-from-date" class="form-control" value="${this.filters.from_date}" readonly>
							</div>
							<div class="col-md-3">
								<label>To Date</label>
								<input type="date" id="donation-to-date" class="form-control" value="${this.filters.to_date}" readonly>
							</div>
						</div>
						<div class="row" style="margin-top: 10px;">
							<div class="col-md-12">
								<button class="btn btn-primary" id="donation-apply-filters">
									<i class="fa fa-filter"></i> Apply
								</button>
								<button class="btn btn-default" id="donation-print">
									<i class="fa fa-print"></i> Print
								</button>
							</div>
						</div>
					</div>
					<div class="donation-report-note">
						Current Period: <strong id="donation-period-label"></strong>
					</div>
					<div class="donation-period-chips" style="margin: 6px 0 10px 0;">
						<button class="btn btn-xs btn-default period-chip" data-period="month">Month</button>
						<button class="btn btn-xs btn-default period-chip" data-period="quarter">Quarter</button>
						<button class="btn btn-xs btn-default period-chip" data-period="half_yearly">Half Yearly</button>
						<button class="btn btn-xs btn-default period-chip" data-period="yearly">Yearly</button>
					</div>
					<div id="donation-report-content"></div>
				`);

				const periodOptions = [
					{ value: "month", label: "Month" },
					{ value: "quarter", label: "Quarter" },
					{ value: "half_yearly", label: "Half Yearly" },
					{ value: "yearly", label: "Yearly" },
				];
				const $period = $("#donation-period-type");
				$period.empty();
				periodOptions.forEach((opt) => {
					$period.append(`<option value="${opt.value}">${opt.label}</option>`);
				});
				$period.val(this.filters.period_type);
				this.render_period_label();
				console.log("[DonationCollection] period dropdown initialized", { selected: $period.val() });
			}

			bind_events() {
				$("#donation-period-type, #donation-reference-date").on("change", () => {
					this.filters.period_type = this.normalize_period_type($("#donation-period-type").val());
					this.filters.reference_date = $("#donation-reference-date").val() || frappe.datetime.get_today();
					console.log("[DonationCollection] filter changed", {
						period_type: this.filters.period_type,
						reference_date: this.filters.reference_date,
					});
					this.set_dates_from_period();
					$("#donation-from-date").val(this.filters.from_date);
					$("#donation-to-date").val(this.filters.to_date);
					this.render_period_label();
					this.sync_period_chips();
				});

				$(".period-chip").on("click", (e) => {
					const period = this.normalize_period_type($(e.currentTarget).data("period"));
					console.log("[DonationCollection] period chip clicked", { period });
					this.filters.period_type = period;
					$("#donation-period-type").val(period);
					this.set_dates_from_period();
					$("#donation-from-date").val(this.filters.from_date);
					$("#donation-to-date").val(this.filters.to_date);
					this.render_period_label();
					this.sync_period_chips();
					this.load_data();
				});

				$("#donation-apply-filters").on("click", () => {
					this.filters.period_type = this.normalize_period_type($("#donation-period-type").val());
					this.filters.reference_date = $("#donation-reference-date").val() || frappe.datetime.get_today();
					this.set_dates_from_period();
					$("#donation-from-date").val(this.filters.from_date);
					$("#donation-to-date").val(this.filters.to_date);
					console.log("[DonationCollection] apply clicked", {
						period_type: this.filters.period_type,
						from_date: this.filters.from_date,
						to_date: this.filters.to_date,
					});
					this.render_period_label();
					this.load_data();
					frappe.show_alert({
						message: `Applied: ${this.pretty_period(this.filters.period_type)} (${this.filters.from_date} to ${this.filters.to_date})`,
						indicator: "green",
					});
				});

				$("#donation-print").on("click", () => window.print());
			}

			render_period_label() {
				console.log("[DonationCollection] render_period_label", {
					period_type: this.filters.period_type,
					from_date: this.filters.from_date,
					to_date: this.filters.to_date,
				});
				$("#donation-period-label").text(
					`${this.pretty_period(this.filters.period_type)} (${this.filters.from_date} to ${this.filters.to_date})`
				);
			}

			sync_period_chips() {
				console.log("[DonationCollection] sync_period_chips", { period_type: this.filters.period_type });
				$(".period-chip").removeClass("btn-primary").addClass("btn-default");
				$(`.period-chip[data-period="${this.filters.period_type}"]`).removeClass("btn-default").addClass("btn-primary");
			}

			pretty_period(periodType) {
				periodType = this.normalize_period_type(periodType);
				const map = {
					month: "Month",
					quarter: "Quarter",
					half_yearly: "Half Yearly",
					yearly: "Yearly",
				};
				return map[periodType] || periodType;
			}

			normalize_period_type(periodType) {
				const value = (periodType || "").toString().trim().toLowerCase();
				if (value === "month" || value === "monthly") return "month";
				if (value === "quarter" || value === "quarterly") return "quarter";
				if (value === "year" || value === "yearly" || value === "annual") return "yearly";
				if (value === "half_yearly" || value === "half-yearly" || value === "half yearly" || value === "halfyearly")
					return "half_yearly";
				return "half_yearly";
			}

			set_dates_from_period() {
				this.filters.period_type = this.normalize_period_type(this.filters.period_type);
				const ref = frappe.datetime.str_to_obj(this.filters.reference_date || frappe.datetime.get_today());
				const year = ref.getFullYear();
				const month = ref.getMonth() + 1;
				let fromDate;
				let toDate;

				if (this.filters.period_type === "month") {
					fromDate = new Date(year, month - 1, 1);
					toDate = new Date(year, month, 0);
				} else if (this.filters.period_type === "quarter") {
					const quarterStartMonth = Math.floor((month - 1) / 3) * 3 + 1;
					fromDate = new Date(year, quarterStartMonth - 1, 1);
					toDate = new Date(year, quarterStartMonth + 2, 0);
				} else if (this.filters.period_type === "yearly") {
					fromDate = new Date(year, 0, 1);
					toDate = new Date(year, 11, 31);
				} else {
					if (month <= 6) {
						fromDate = new Date(year, 0, 1);
						toDate = new Date(year, 5, 30);
					} else {
						fromDate = new Date(year, 6, 1);
						toDate = new Date(year, 11, 31);
					}
				}

				this.filters.from_date = frappe.datetime.obj_to_str(fromDate);
				this.filters.to_date = frappe.datetime.obj_to_str(toDate);
				console.log("[DonationCollection] set_dates_from_period computed", {
					period_type: this.filters.period_type,
					reference_date: this.filters.reference_date,
					from_date: this.filters.from_date,
					to_date: this.filters.to_date,
				});
			}

			load_data() {
				console.log("[DonationCollection] load_data() start", {
					from_date: this.filters.from_date,
					to_date: this.filters.to_date,
				});
				frappe.call({
					method:
						"tif_customization.tif_customization.page.donation_collection_.donation_collection_.get_donation_collection_data",
					args: {
						filters: {
							period_type: this.filters.period_type,
							reference_date: this.filters.reference_date,
							from_date: this.filters.from_date,
							to_date: this.filters.to_date,
						},
					},
					callback: (r) => {
						const message = r.message || {};
						const months = message.months || [];
						const rows = message.rows || [];
						const totals = message.totals || {};
						console.log("[DonationCollection] load_data() success", {
							row_count: rows.length,
							months_count: months.length,
							totals,
						});
						this.render_data_table(months, rows, totals);
					},
					error: (err) => {
						console.error("[DonationCollection] load_data() failed", err);
						$("#donation-report-content").html(
							`<div class="text-danger">Failed to load donation collection data.</div>`
						);
					},
				});
			}

			render_data_table(months, rows, totals) {
				if (!rows.length) {
					$("#donation-report-content").html(
						`<div class="text-muted" style="margin-top:10px;">No donation records found for selected period.</div>`
					);
					return;
				}

				const monthHead = months
					.map((m) => `<th class="text-right">${frappe.utils.escape_html(m.label)}</th>`)
					.join("");
				const periodBandLabel = `${this.pretty_period(this.filters.period_type)} (${this.filters.from_date} to ${this.filters.to_date})`;

				const body = rows
					.map(
						(row) => `
						<tr>
							<td>${frappe.utils.escape_html(row.donor_name || "")}</td>
							<td>${frappe.utils.escape_html(row.donor_id || "-")}</td>
							${months
								.map((m) => `<td class="text-right">${this.format_currency((row.month_values || {})[m.key])}</td>`)
								.join("")}
							<td class="text-right summary-start">${this.format_currency(row.committed_amount)}</td>
							<td class="text-right">${this.format_currency(row.other_collection)}</td>
							<td class="text-right">${this.format_currency(row.zakat_fund)}</td>
							<td class="text-right">${this.format_currency(row.total_collection)}</td>
							<td class="text-right">${this.format_currency(row.balance_amount)}</td>
						</tr>
					`
					)
					.join("");

				$("#donation-report-content").html(`
					<div class="table-responsive" style="margin-top: 10px;">
						<table class="table table-bordered donation-matrix-table">
							<thead>
								<tr>
									<th rowspan="2">Donor</th>
									<th rowspan="2">Donor ID</th>
									<th colspan="${months.length}" class="period-band">${frappe.utils.escape_html(periodBandLabel)}</th>
									<th colspan="5" class="summary-band summary-start">Summary</th>
								</tr>
								<tr>
									${monthHead}
									<th class="text-right">Commitment</th>
									<th class="text-right">Donation/Other Collection</th>
									<th class="text-right">Zakat Fund</th>
									<th class="text-right">Total Received</th>
									<th class="text-right">Balance</th>
								</tr>
							</thead>
							<tbody>
								${body}
								<tr class="grand-total-row">
									<td colspan="2">Grand Total</td>
									${months
										.map((m) => `<td class="text-right">${this.format_currency((totals.month_values || {})[m.key])}</td>`)
										.join("")}
									<td class="text-right summary-start">${this.format_currency(totals.committed_amount)}</td>
									<td class="text-right">${this.format_currency(totals.other_collection)}</td>
									<td class="text-right">${this.format_currency(totals.zakat_fund)}</td>
									<td class="text-right">${this.format_currency(totals.total_collection)}</td>
									<td class="text-right">${this.format_currency(totals.balance_amount)}</td>
								</tr>
							</tbody>
						</table>
					</div>
				`);
			}

			format_currency(v) {
				return format_currency(v || 0, frappe.defaults.get_default("currency"));
			}

			inject_styles() {
				if ($("#donation-matrix-style").length) return;
				$("head").append(`
					<style id="donation-matrix-style">
						.donation-matrix-table thead th {
							font-weight: 700;
							vertical-align: middle;
						}
						.donation-matrix-table thead .period-band {
							background: #cfd6e2;
							text-align: center;
							font-size: 15px;
						}
						.donation-matrix-table thead .summary-band {
							background: #dde5f0;
							text-align: center;
						}
						.donation-matrix-table tbody tr:nth-child(odd) {
							background: #fbfcfe;
						}
						.donation-matrix-table .summary-start {
							border-left: 3px solid #c1c9d6 !important;
						}
						.donation-matrix-table .grand-total-row td {
							font-weight: 700;
							background: #d7deea;
						}
					</style>
				`);
			}
		};
	}

	const report = new window.DonationCollectionReport(page);
	report.make();
	report.sync_period_chips();
	console.log("[DonationCollection] report initialized");
};
